public/app/plugins/datasource/loki/querybuilder/components/NestedQuery.test.tsx

```diff
@@ -0,0 +1,112 @@
+import { fireEvent, render, screen, waitFor } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import React from 'react';
+
+import { DataSourceInstanceSettings, DataSourcePluginMeta } from '@grafana/data';
+
+import { LokiDatasource } from '../../datasource';
+import { LokiVisualQueryBinary } from '../types';
+
+import { NestedQuery, Props as NestedQueryProps } from './NestedQuery';
+
+type Operator = '+' | '-' | '*' | '/' | '%' | '^' | '==' | '!=' | '>' | '<' | '>=' | '<=';
+type VectorMatchType = 'on' | 'ignoring';
+
+const createMockProps = (
+  operator: Operator = '/',
+  vectorMatchesType: VectorMatchType = 'on',
+  showExplain = false
+): NestedQueryProps => {
+  const nestedQuery: LokiVisualQueryBinary = {
+    operator: operator,
+    vectorMatchesType: vectorMatchesType,
+    query: {
+      labels: [],
+      operations: [],
+    },
+  };
+
+  const datasource = new LokiDatasource(
+    {
+      url: '',
+      jsonData: {},
+      meta: {} as DataSourcePluginMeta,
+    } as DataSourceInstanceSettings,
+    undefined,
+    undefined
+  );
+
+  const props: NestedQueryProps = {
+    nestedQuery: nestedQuery,
+    index: 0,
+    datasource: datasource,
+    onChange: jest.fn(),
+    onRemove: jest.fn(),
+    onRunQuery: jest.fn(),
+    showExplain: showExplain,
+  };
+
+  return props;
+};
+
+// All test assertions need to be awaited for, because the component uses `useEffect` to update the state.
```

Comment:
I don't see any effect or state in the component. What is this comment about?

Reply:
I checked and it's actually `LokiQueryBuilder` the one triggering the async updates. No need to change anything, but sometimes a good approach to take care of async dependencies is mocking them, if you don't need to interact with them.

E.g.

```jsx
jest.mock('./LokiQueryBuilder', () => () => <div data-testid="loki-query-builder-mock" />);

// ...
expect(screen.getByTestId('loki-query-builder-mock')).toBeInTheDocument();
```

Reply:
just to make sure i'm understanding this.

when using this, if i `render(<NestedQuery {...props} />);`, `<LokiQueryBuilder />` wont actually be rendered inside of NestedQuery instead it will render `<div data-testid="loki-query-builder-mock" />` as a mock.?

Reply:
Correct. 

Again, no need change anything here, just pointing out it's a possibility when you're writing tests.